File: app/src/main/java/com/iceteaviet/fastfoodfinder/model/routing/Leg.java
package com.iceteaviet.fastfoodfinder.model.routing;

import android.os.Parcel;
import android.os.Parcelable;

import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.google.gson.annotations.SerializedName;

import java.util.List;

/**
 * Created by Genius Doan on 11/11/2016.
 */
public class Leg implements Parcelable {

    public static final Creator<Leg> CREATOR = new Creator<Leg>() {
        @Override
        public Leg createFromParcel(Parcel in) {
            return new Leg(in);
        }

        @Override
        public Leg[] newArray(int size) {
            return new Leg[size];
        }
    };
    @SerializedName("distance")
    JsonObject distance;
    @SerializedName("duration")
    JsonObject duration;
    @SerializedName("start_address")
    String startAddress;
    @SerializedName("end_address")
    String endAddress;
    @SerializedName("steps")
    List<Step> stepList;

    protected Leg(Parcel in) {
        JsonParser parser = new JsonParser();
        distance = parser.parse(in.readString()).getAsJsonObject();
        duration = parser.parse(in.readString()).getAsJsonObject();
        startAddress = in.readString();
        endAddress = in.readString();
        stepList = in.createTypedArrayList(Step.CREATOR);
    }

    public String getDistance() {
        return distance.getAsJsonPrimitive("text").getAsString();
    }

    public long getDistanceValue() {
        return distance.getAsJsonPrimitive("value").getAsLong();
    }

    public String getDuration() {
        return duration.getAsJsonPrimitive("text").getAsString();
    }

    public long getDurationValue() {
        return duration.getAsJsonPrimitive("value").getAsLong();
    }

    public String getStartAddress() {
        return startAddress;
    }

    public String getEndAddress() {
        return endAddress;
    }

    public List<Step> getStepList() {
        return stepList;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel parcel, int i) {
        parcel.writeString(distance.toString());
        parcel.writeString(duration.toString());
        parcel.writeString(startAddress);
        parcel.writeString(endAddress);
        parcel.writeTypedList(stepList);
    }
}

File: app/src/main/kotlin/com/iceteaviet/fastfoodfinder/data/remote/routing/model/Leg.kt
package com.iceteaviet.fastfoodfinder.data.remote.routing.model

import android.os.Parcel
import android.os.Parcelable

import com.google.gson.JsonObject
import com.google.gson.JsonParser
import com.google.gson.annotations.SerializedName

/**
 * Created by Genius Doan on 11/11/2016.
 */
class Leg constructor(`in`: Parcel) : Parcelable {
    @SerializedName("distance")
    private val distance: JsonObject
    @SerializedName("duration")
    private val duration: JsonObject
    @SerializedName("start_address")
    val startAddress: String
    @SerializedName("end_address")
    val endAddress: String
    @SerializedName("steps")
    val stepList: List<Step>

    val distanceValue: Long
        get() = distance.getAsJsonPrimitive("value").asLong

    val durationValue: Long
        get() = duration.getAsJsonPrimitive("value").asLong

    init {
        val parser = JsonParser()
        distance = parser.parse(`in`.readString()).asJsonObject
        duration = parser.parse(`in`.readString()).asJsonObject
        startAddress = `in`.readString() ?: ""
        endAddress = `in`.readString() ?: ""
        stepList = `in`.createTypedArrayList(Step.CREATOR) ?: ArrayList()
    }

    fun getDistance(): String {
        return distance.getAsJsonPrimitive("text").asString
    }

    fun getDuration(): String {
        return duration.getAsJsonPrimitive("text").asString
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(parcel: Parcel, i: Int) {
        parcel.writeString(distance.toString())
        parcel.writeString(duration.toString())
        parcel.writeString(startAddress)
        parcel.writeString(endAddress)
        parcel.writeTypedList(stepList)
    }

    companion object CREATOR : Parcelable.Creator<Leg> {
        override fun createFromParcel(parcel: Parcel): Leg {
            return Leg(parcel)
        }

        override fun newArray(size: Int): Array<Leg?> {
            return arrayOfNulls(size)
        }
    }
}
